src/asgen/backends/alpinelinux/apkpkgindex.d

```diff
@@ -0,0 +1,202 @@
+ * (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this software.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+module asgen.backends.alpinelinux.apkpkgindex;
+
+import std.algorithm : canFind, filter, endsWith;
+import std.array : appender, split;
+import std.conv : to;
+import std.exception : enforce;
+import std.file : dirEntries, exists, SpanMode;
+import std.format : format;
+import std.path : baseName, buildPath;
+import std.string : splitLines, startsWith, strip;
+import std.utf : UTFException, validate;
+
+import asgen.logging;
+import asgen.zarchive;
+import asgen.utils : escapeXml;
+import asgen.backends.interfaces;
+import asgen.backends.alpinelinux.apkpkg;
+
+final class AlpinePackageIndex : PackageIndex
+{
+
+private:
+    string rootDir;
+    Package[][string] pkgCache;
+
+public:
+
+    this(string dir)
+    {
+        enforce(exists(dir), format("Directory '%s' does not exist.", dir));
+        this.rootDir = dir;
+    }
+
+    override void release()
+    {
+        pkgCache = null;
+    }
+
+    private void setPkgDescription(AlpinePackage pkg, string pkgDesc)
+    {
+        if (pkgDesc is null)
+        {
+            return;
+        }
+
+        auto desc = "<p>%s</p>".format(pkgDesc.escapeXml);
+        pkg.setDescription(desc, "C");
+    }
+
+    private Package[] loadPackages(string suite, string section, string arch)
+    {
+        auto apkRootPath = buildPath(rootDir, suite, section, arch);
+        ArchiveDecompressor ad;
+        AlpinePackage[string] pkgsMap;
+
+        foreach (packageArchivePath; dirEntries(apkRootPath, SpanMode.shallow).filter!(
+                f => f.name.endsWith(".apk")))
+        {
+            auto fileName = packageArchivePath.baseName();
+            AlpinePackage pkg;
+            if (fileName in pkgsMap)
+            {
+                pkg = pkgsMap[fileName];
+            }
+            else
+            {
+                pkg = new AlpinePackage();
+                pkgsMap[fileName] = pkg;
+            }
+
+            ad.open(packageArchivePath);
+            auto pkgInfoData = cast(string) ad.readData(".PKGINFO");
+
+            try
+            {
+                validate(pkgInfoData);
+            }
+            catch (UTFException e)
+            {
+                logError("PKGINFO file in archive %s contained invalid UTF-8, skipping!",
+                        packageArchivePath);
+            }
+
+            pkg.filename = packageArchivePath;
+            auto lines = pkgInfoData.splitLines();
+            for (auto i = 0; i < lines.length; i++)
+            {
+                /* Skip comments */
```

Comment:
style: Don't use C-style comment blocks unless needed, and don't mix comment styles (there's multiple styles of comments in this file)